### PR TITLE
Adding Input component and updating TextFieldBase

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## Explanation
Adding `Input` component as a base component for `TextFieldBase`. Based on [TextField Audit and Insight Report](https://www.figma.com/file/aGW8sk6X6Jf9ac0MRMD4kX/TextField-Audit?node-id=349%3A23&t=xHFfsVJIbeO0FOxP-1)

Will be updating more TextField related components in other PRs for brevity 

#17584 

## Screenshots/Screencaps

https://user-images.githubusercontent.com/8112138/217631063-ca46d708-5509-4359-8e16-630870044dc3.mov

## Manual Testing Steps

- Go to the latest storybook build of this PR
- Search `Input` in the search panel
- Go to `Input` story in **Component Library** 
- Check all controls work
- Check all stories pass accessibility 
- Check documentation in Docs tabs
- Check console for errors

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

### CODE REVIEW

File: ui/components/component-library/input/README.mdx [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+Use the `type` prop to change the type of input.
+
+Possible types include:
+
+- `INPUT_TYPES.TEXT`
+- `INPUT_TYPES.NUMBER`
+- `INPUT_TYPES.PASSWORD`
+- `INPUT_TYPES.SEARCH`
+
+Defaults to `text`.
```
Comment (georgewrmarshall) [inline ui/components/component-library/input/README.mdx:30]: Update to 
```suggestion
Defaults to `INPUT_TYPES.TEX`.
```

File: ui/components/component-library/input/README.mdx [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+```jsx
+import { Input } from '../../component-library';
+
+<Input autoFocus />;
+```
+
+### Default Value
+
+Use the `defaultValue` prop to set the default value of the `Input`
```
Comment (georgewrmarshall) [inline ui/components/component-library/input/README.mdx:105]: Update to 
```suggestion
Use the `defaultValue` prop to set the default value of the `Input`. Use for uncontrolled inputs
```

File: ui/components/component-library/input/README.mdx [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+```jsx
+import { Input } from '../../component-library';
+
+<Input disabled />;
+```
+
+### Error
+
+Use the `error` prop to set `aria-invalid="true"`. This helps with screen readers for accessibility.
```
Comment (georgewrmarshall) [inline ui/components/component-library/input/README.mdx:133]: Update to
```suggestion
Use the `error` prop to set `aria-invalid="true"`. This helps with screen readers for accessibility. There is no visual indicator for `error` this should be handled in the parent component.
```

File: ui/components/component-library/input/README.mdx [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+```jsx
+import { Input } from '../../component-library';
+
+<Input maxLength={10} />;
+```
+
+### Read Only
+
+Use the `readOnly` prop to set the `Input` to read only. When `readOnly` is true `Input` will not have a focus state.
```
Comment (georgewrmarshall) [inline ui/components/component-library/input/README.mdx:161]: Update to
```suggestion
Use the `readOnly` prop to set the `Input` to read only
```

File: ui/components/component-library/input/README.mdx [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+```jsx
+import { Input } from '../../component-library';
+
+<Input readOnly />;
+```
+
+### Required
+
+Use the `required` prop to set the html `required` attribute. There is no visual difference to the `Input` when required is true but is used by the browser api.
```
Comment (georgewrmarshall) [inline ui/components/component-library/input/README.mdx:175]: Update to
```suggestion
Use the `required` prop to set the html `required` attribute used by the browser api. There is no visual indicator for `required` this should be handled in the parent component.  
```

File: ui/components/component-library/input/README.mdx [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+<Canvas>
+  <Story id="components-componentlibrary-input--required" />
+</Canvas>
+
+```jsx
+import { Input } from '../../component-library';
+
+// No visual difference. Used by the browser api
```
Comment (georgewrmarshall) [inline ui/components/component-library/input/README.mdx:184]:
```suggestion
// No visual indicator. Used by the browser api
```

File: ui/components/component-library/text-field-base/text-field-base.scss [hunk 7]
```diff
@@ -38,17 +38,7 @@
...
     width: 100%;
     flex-grow: 1;
-    box-sizing: content-box;
-    margin: 0;
-    padding: 0;
-
-    &:focus,
-    &:focus-visible {
-      outline: none;
-    }
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field-base/text-field-base.scss:52]: Removing styles that are handled by the `input.scss`

File: ui/components/component-library/text-field-base/text-field-base.stories.js [hunk 8]
```diff
@@ -151,11 +150,13 @@ export default {
   },
   args: {
     placeholder: 'Placeholder...',
+    value: '',
   },
 };
 
 const Template = (args) => {
   const [{ value }, updateArgs] = useArgs();
+  console.log('value', value);
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field-base/text-field-base.stories.js:159]:
```suggestion
```

File: ui/components/component-library/text-field-base/text-field-base.stories.js [hunk 9]
```diff
@@ -212,6 +213,10 @@ export const Type = (args) => (
   </Box>
 );
 
+Type.args = {
+  value: undefined,
+};
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field-base/text-field-base.stories.js:219]: Fixes and error in storybook swapping from controlled vs uncontrolled inputs

File: ui/components/component-library/text-field-base/text-field-base.stories.js [hunk 10]
```diff
@@ -484,8 +489,9 @@ AutoComplete.args = {
 export const AutoFocus = Template.bind({});
 AutoFocus.args = { autoFocus: true };
 
-export const DefaultValue = Template.bind({});
-DefaultValue.args = { defaultValue: 'Default value' };
+export const DefaultValue = () => (
+  <TextFieldBase placeholder="Default value" defaultValue="Default value" />
+);
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field-base/text-field-base.stories.js:494]: Fixes error in storybook swapping between uncontrolled vs controlled inputs

**Comment by georgewrmarshall** [comment]
> `TextFieldBase` props aren't show up in `TextField`
> 
> Screenshot 2023-02-09 at 1 33 42 PM

Looks like it was already broken going to address it in a separate PR https://github.com/MetaMask/metamask-extension/issues/17723

File: ui/components/component-library/input/README.mdx [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+<Input type={INPUT_TYPES.PASSWORD} autoComplete />;
+```
+
+### Auto Focus
+
+Use the `autoFocus` prop to focus the `Input` during the first mount
+
+<Canvas>
+  <Story id="components-componentlibrary-input--auto-focus" />
```
[the remark this answers is not in the document]
Comment (Sagi111) [reply ui/components/component-library/input/README.mdx:94]: .github/workflows/blank.yml

[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/input/README.mdx]: Yeah good point. It's very annoying. Removed here: https://github.com/MetaMask/metamask-extension/pull/17664/commits/6a1e1b6fe5938d058f48ae933139f3874a79680f


https://user-images.githubusercontent.com/8112138/218845047-9274ef28-4175-48ce-9e84-1f5a70e6c14f.mov

**Review by garrettbear — DISMISSED** [review]

**Comment by georgewrmarshall** [comment]
I think it's out of scope of this ticket but if we needed a textarea immediately you could use the polymorphic `as` prop to change the root element `<Input as="textarea" />` and it would work but there isn't any multiline logic built in. We can make that decision when we get to auditing the TextArea

File: ui/components/component-library/text-field/text-field.js [hunk 12]
```diff
@@ -115,7 +114,7 @@ export const TextField = ({
       <InputComponent
         aria-invalid={error}
         as="input"
-        autoComplete={autoComplete ? 'on' : 'off'}
+        autoComplete={autoComplete}
```
Comment (NidhiKJha) [inline ui/components/component-library/text-field/text-field.js:117]: Since, InputComponent = Input already, do we need `as="input"` here?

**Review by NidhiKJha — COMMENTED** [review]
Looking great! 🚀  Just one small nit. Also, need to export input in `ui/components/component-library/index.js`

`export { Input, INPUT_TYPES } from './input';`

File: ui/components/component-library/text-field/text-field.js (hunk 12 above)
Comment (georgewrmarshall) [reply ui/components/component-library/text-field/text-field.js]: Yes! Resolved https://github.com/MetaMask/metamask-extension/pull/17664/commits/f3e19eea5648363b740c565208518b4ab8c8225b

**Review by garrettbear — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]

**Pull request merged**